### PR TITLE
sql/stats: read the list of initial tables asynchronously

### PR DESCRIPTION
**Pull request by knz**
Recommended by @RaduBerinde and @bdarnell in #35549 

This patch changes the refresher init code to perform the initial scan
through all descriptors using ASOT, to avoid possible txn conflicts
with the first client connections (or tests).

Release note: None

### CODE REVIEW

**Review by RaduBerinde — APPROVED** [review]
:lgtm:  Does this help with the issue you were seeing?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35552#-:-L_TVeXIBTSGYbMaVFmJ:b-qdbiz5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

**Comment by bdarnell** [comment]
I would prefer to keep this asynchronous, but to use AOST to avoid disrupting other clients (in a multi-node cluster, each of them will independently be performing this step, so it could be run while another node is already serving its sql clients). That probably means adding something like a sleep(5s) at the start of the worker and then using AOST -5s when doing the read.

**Comment by knz** [comment]
Added another commit to use ASOT as recommended by @bdarnell in https://github.com/cockroachdb/cockroach/issues/35549#issuecomment-471007721

**Comment by knz** [comment]
RFAL

**Review by RaduBerinde — COMMENTED** [review]
CC @rytaft 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35552#-:-L_TYoAb74nWHtSUqomS:b-y9cmbr)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz and @rytaft)

___
*[pkg/sql/stats/automatic_stats.go, line 283 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35552#-L_TY0Ai3D0I6jDYtsm9:-L_TY0Ai3D0I6jDYtsmA:bbapv3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6d4acce503b48e9bc76a3782c0f173f22adfb8a/pkg/sql/stats/automatic_stats.go#L283)):*
> ```Go
> 	asot := ""
> 	if !firstNode {
> 		// On every node but the first, we use a historical read so as to
> ```

I don't quite get why we treat nodes differently. I think if we move this into the worker and sleep for a bit we can use AOST on all of them.

I would make sure that AOST indeed gets used internally (this is a virtual table which internally issues another query).

**Comment by knz** [comment]
> I don't quite get why we treat nodes differently. I think if we move this into the worker and sleep for a bit we can use AOST on all of them.

Done. RFAL

(We have logic tests that confirm that AOST applies to virtual tables already so that should be fine.)

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35552)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz and @rytaft)

___
*[pkg/sql/stats/automatic_stats.go, line 270 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/35552#-L_TcU9a7NQtSAOkUJTy:-L_TcU9a7NQtSAOkUJTz:b-huv5y5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0bde492eff6533123f56fa0f4428b0559415d195/pkg/sql/stats/automatic_stats.go#L270)):*
> ```Go
> 	// We wait for the asOfTime delay and some more before initialization,
> 	// to ensure that the historical read will pick up all system tables.
> 	time.Sleep(r.asOfTime + time.Second)
> ```

asOfTime is 30 seconds by default. I am a bit worried that we might hit the "buffered channel is full" case in `NotifyMutation` by the time this worker up and running. I would just sleep for 1 second (maybe a tad more) and use AOST -1s.

**Comment by knz** [comment]
> I  would just sleep for 1 second (maybe a tad more) and use AOST -1s.

Done.

File: pkg/sql/stats/automatic_stats.go [hunk 1]
```diff
@@ -265,11 +265,15 @@ func (r *Refresher) ensureAllTables(ctx context.Context, settings *settings.Valu
...
+	// We wait for the asOfTime delay and some more before initialization,
+	// to ensure that the historical read will pick up all system tables.
+	time.Sleep(time.Second)
+
 	rows, err := r.ex.Query(
 		ctx,
 		"get-tables",
 		nil, /* txn */
-		`SELECT table_id FROM crdb_internal.tables;`,
+		fmt.Sprintf(`SELECT table_id FROM crdb_internal.tables AS OF SYSTEM TIME '-%s'`, time.Microsecond),
```
Comment (RaduBerinde) [inline pkg/sql/stats/automatic_stats.go]: why 1 microsecond? Also I think sprintf is overkill given the value is hardcoded

Comment (knz) [reply pkg/sql/stats/automatic_stats.go]: Added a comment to clarify. Good call on the constant, simplified.

**Review by bdarnell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35552)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @RaduBerinde, and @rytaft)

___
*[pkg/sql/stats/automatic_stats.go, line 270 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/35552#-L_UMt8xEthD3Fq0LIk_:-L_UMt8xEthD3Fq0LIka:b52pkdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c660e45c732d207e2c002dca6b857d9b32fe8f39/pkg/sql/stats/automatic_stats.go#L270)):*
> ```Go
> 	// We wait for some time before initialization, to ensure that the
> 	// historical read below will pick up all system tables.
> 	time.Sleep(time.Second)
> ```

I think this will slow down the tests since time.Sleep is uninterruptible, so we'll impose a minimum duration of 1s on any test that includes this code. Instead, use time.After in a select with the context cancellation channel.
___
*[pkg/sql/stats/automatic_stats.go, line 276 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/35552#-L_UM2DcDCHpb1AUO9sy:-L_UM2DcDCHpb1AUO9sz:bnkziqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/c660e45c732d207e2c002dca6b857d9b32fe8f39/pkg/sql/stats/automatic_stats.go#L276)):*
> ```Go
> 	// 1s delay we just waited for above, so that we don't
> 	// forget any newly created tables.
> 	const getAllTablesQuery = `SELECT table_id FROM crdb_internal.tables AS OF SYSTEM TIME '-1us'`
> ```

I think 1µs is too small - we'll still be at risk of causing serializability restarts in other transactions. 1µs in the past isn't much different from running at the current time. (it avoids uncertainty errors but that's not our concern here).

The idea behind the sleep and AOST combination is that before we sleep, we install our hook to watch for new tables as they are created (I assume we must have such a thing if we only do the all-table scan at startup). That way we'll find out about new tables as they happen, and 1s after startup we'll do a backfill of tables that existed previously. This can run at AOST -1s because we know we've heard about newer tables in other ways.

**Comment by knz** [comment]
Thanks Ben for the hint about doing the right thing with the timer channel. I re-organized the code accordingly. PTAL.

**Review by RaduBerinde — APPROVED** [review]
This latest iteration looks great. Thank you for working through this. :lgtm: 

CC @rytaft - just so you're aware of the change.

> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35552#-:-L_V4QjFDRqPG8x3ey7x:bgq0kys)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @rytaft)

**Review by bdarnell — APPROVED** [review]
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35552)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @rytaft)

**Comment by knz** [comment]
Thanks!

bors r+

**Comment by rytaft** [comment]
Thanks for doing this work!

**Pull request merged**